File: leetcode/1917-maximum-average-pass-ratio/solution.go
type Class struct {
	Bonus float64
	Passed int
	Total  int
}

type MaxHeap []Class

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i].Bonus > h[j].Bonus }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) { *h = append(*h, x.(Class)) }
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	item := old[n-1]
	*h = old[:n-1]
	return item
}

func calculateBonus(passed, total int) float64 {
	p := float64(passed)
	n := float64(total)
	return (n - p) / n / (n + 1.0)
}

func maxAverageRatio(classes [][]int, extraStudents int) float64 {
	maxHeap := &MaxHeap{}
	heap.Init(maxHeap)

	// Populate the heap
	for _, class := range classes {
		passed, total := class[0], class[1]
		heap.Push(maxHeap, Class{
			Bonus:  calculateBonus(passed, total),
			Passed: passed,
			Total:  total,
		})
	}

	// Distribute extra students
	for extraStudents > 0 {
		class := heap.Pop(maxHeap).(Class)
		class.Passed++
		class.Total++
		class.Bonus = calculateBonus(class.Passed, class.Total)
		heap.Push(maxHeap, class)
		extraStudents--
	}

	// Calculate final average ratio
	totalRatio := 0.0
	for maxHeap.Len() > 0 {
		class := heap.Pop(maxHeap).(Class)
		totalRatio += float64(class.Passed) / float64(class.Total)
	}

	return totalRatio / float64(len(classes))
}

File: leetcode/1917-maximum-average-pass-ratio/solution.rs
impl Solution {
    pub fn max_average_ratio(classes: Vec<Vec<i32>>, mut extra_students: i32) -> f64 {
        use std::collections::BinaryHeap;
        let K = 1000_000_000.0;
        let mut classes: BinaryHeap<(i64, i32, i32)> = classes.into_iter()
            .map(|a| (a[0] as f64, a[1] as f64))
            .map(|(p, n)| {
                (((n-p)/n/(n+1.0)*K) as i64, p as i32, n as i32)
            })
            .collect();
        while extra_students > 0 {
            let (_, mut p, mut n) = classes.pop().unwrap();
            p += 1;
            n += 1;
            let p = p as f64;
            let n = n as f64;
            let bonus = ((n-p)/n/(n+1.0)*K) as i64;
            classes.push((bonus, p as i32, n as i32));
            extra_students -= 1;
        }
        let n = classes.len() as f64;
        classes.into_iter().map(|(_, p, n)| p as f64/n as f64).sum::<f64>() / n
    }
}
